lib/index/response_validator.rb

```diff
@@ -0,0 +1,61 @@
+module Index
+  class ResponseValidator
+    class NotFound < StandardError; end
+    class ElasticsearchError < StandardError; end
+
+    def self.valid!(response)
+      action_type, details = response.first # response is a hash with a single [key, value] pair
+      status = details['status']
+
+      if (200..399).cover?(status)
+        logger.debug("Processed #{action_type} with status #{status}")
+        Services.statsd_client.increment("metasearch.elasticsearch.#{action_type}")
+      elsif status == 404 # failed while attempting to delete missing record so just ignore it
+        logger.info("Tried to delete a document that wasn't there; ignoring.")
+        Services.statsd_client.increment('metasearch.elasticsearch.already_deleted')
+        raise NotFound, "Document not found in index"
+      else
+        logger.error("#{action_type} not processed: status #{status}")
+        Services.statsd_client.increment("metasearch.elasticsearch.#{action_type}_error")
+
+        raise ElasticsearchError, "Unknown Error"
+      end
+    end
+
+    def self.valid?(response)
+      action_type, details = response.first # response is a hash with a single [key, value] pair
+      status = details['status']
+
+      if (200..399).cover?(status)
+        logger.debug("Processed #{action_type} with status #{status}")
+        Services.statsd_client.increment("govuk_index.elasticsearch.#{action_type}")
```

Comment:
do we want to parameterise the statsd key name? It always includes "govuk_index" at the moment.

Reply:
good point.. i'll think of a way to fix this.